File: src/main/scala/com/snowflake/snowpark/types/Geometry.scala
package com.snowflake.snowpark.types

import java.io.{IOException, UncheckedIOException}

/** Companion object of Geometry class.
  * @since 1.12.0
  */
object Geometry {

  /** Creates a Geometry class from a GeoJSON string
    *
    * @param g
    *   GeoJSON string
    * @return
    *   a Geometry class
    * @since 1.12.0
    */
  def fromGeoJSON(g: String): Geometry = new Geometry(g)
}

/** Scala representation of Snowflake Geometry data. Only support GeoJSON format.
  *
  * @since 1.12.0
  */
class Geometry private (private val stringData: String) {
  if (stringData == null) {
    throw new UncheckedIOException(new IOException("Cannot create geometry object from null input"))
  }

  /** Returns whether the Geometry object equals to the input object.
    *
    * @return
    *   GeoJSON string
    * @since 1.12.0
    */
  override def equals(obj: Any): Boolean =
    obj match {
      case g: Geometry => stringData.equals(g.stringData)
      case _ => false
    }

  /** Returns the hashCode of the stored GeoJSON string.
    *
    * @return
    *   hash code
    * @since 1.12.0
    */
  override def hashCode(): Int = stringData.hashCode

  /** Returns the underling string data for GeoJSON.
    *
    * @return
    *   GeoJSON string
    * @since 1.12.0
    */
  override def toString: String = stringData
}

File: src/main/java/com/snowflake/snowpark_java/types/Geometry.java
package com.snowflake.snowpark_java.types;

import java.io.IOException;
import java.io.Serializable;
import java.io.UncheckedIOException;

/**
 * Java representation of Snowflake Geometry data.
 *
 * @since 1.12.0
 */
public class Geometry implements Serializable {
  private final String data;

  private Geometry(String data) {
    if (data == null) {
      throw new UncheckedIOException(
          new IOException("Cannot create geometry object from null input"));
    }
    this.data = data;
  }

  /**
   * Checks whether two Geometry object are equal.
   *
   * @param other A Geometry object
   * @return true if these two object are equal
   * @since 1.12.0
   */
  @Override
  public boolean equals(Object other) {
    if (other instanceof Geometry) {
      return data.equals(((Geometry) other).data);
    }
    return false;
  }

  /**
   * Calculates the hash code of this Geometry Object.
   *
   * @return An int number representing the hash code value
   * @since 1.12.0
   */
  @Override
  public int hashCode() {
    return this.data.hashCode();
  }

  /**
   * Converts this Geometry object to a String value.
   *
   * @return A String value.
   * @since 1.12.0
   */
  @Override
  public String toString() {
    return this.data;
  }

  /**
   * Creates a Geometry object from a GeoJSON string.
   *
   * @param g GeoJSON String
   * @return a new Geometry object
   * @since 1.12.0
   */
  public static Geometry fromGeoJSON(String g) {
    return new Geometry(g);
  }
}
